### ISSUE TITLE
Pre-provisioned volumes is not deleted with persistentVolumeReclaimPolicy Delete

### ISSUE THREAD

**Issue by dkoshkin** [kubernetes-sigs/aws-ebs-csi-driver#164]
/kind bug

**What happened?**
Created a `PersistentVolume` resource with an EBS volume and `persistentVolumeReclaimPolicy: Delete`, a `PersistentVolumeClaim` and a matching `Pod`. The volume binds correctly and can write to it.
However, when deleting the resources the EBS does not get deleted.

**What you expected to happen?**
The EBS volume to get deleted. Reading the Kubernetes code thats what it looks like its supposed to do but I'm not 100% certain if that also works for CSI plugins.

**How to reproduce it (as minimally and precisely as possible)?**
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-test2
spec:
  accessModes:
    - ReadWriteOnce
  capacity:
    storage: 1Gi
  csi:
    driver: ebs.csi.aws.com
    fsType: ext3
    volumeHandle: vol-004ccedade8d11dc3
  persistentVolumeReclaimPolicy: Delete
  volumeMode: Filesystem
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: claim1
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: ""
  resources:
    requests:
      storage: 1Gi
---
apiVersion: v1
kind: Pod
metadata:
  name: app
spec:
  containers:
  - name: app
    image: centos:7
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo $(date -u) >> /data/out.txt; sleep 5; done"]
    volumeMounts:
    - name: persistent-storage
      mountPath: /data
  volumes:
  - name: persistent-storage
    persistentVolumeClaim:
      claimName: claim1
```

**Anything else we need to know?**:
controller-manager logs:
```
I1229 01:22:36.423457       1 operation_generator.go:419] DetachVolume.Detach succeeded for volume "pv-test2" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-004ccedade8d11dc3") on node "kube-node-0-kubelet.devkubernetes01.mesos"
E1229 01:24:08.102654       1 goroutinemap.go:150] Operation for "delete-pvc-test[8bf04e73-0ac6-11e9-9b24-16e3c8128b03]" failed. No retries permitted until 2018-12-29 01:26:10.10263137 +0000 UTC m=+39025.756049180 (durationBeforeRetry 2m2s). Error: "Error getting deleter volume plugin for volume \"pvc-test\": no volume plugin matched"
E1229 01:26:23.104891       1 goroutinemap.go:150] Operation for "delete-pvc-test[8bf04e73-0ac6-11e9-9b24-16e3c8128b03]" failed. No retries permitted until 2018-12-29 01:28:25.1048646 +0000 UTC m=+39160.758282509 (durationBeforeRetry 2m2s). Error: "Error getting deleter volume plugin for volume \"pvc-test\": no volume plugin matched"
E1229 01:28:38.106519       1 goroutinemap.go:150] Operation for "delete-pvc-test[8bf04e73-0ac6-11e9-9b24-16e3c8128b03]" failed. No retries permitted until 2018-12-29 01:30:40.106496961 +0000 UTC m=+39295.759914765 (durationBeforeRetry 2m2s). Error: "Error getting deleter volume plugin for volume \"pvc-test\": no volume plugin matched"
```

Looks like the controller cannot find the plugin https://github.com/kubernetes/kubernetes/blob/v1.13.0/pkg/controller/volume/persistentvolume/pv_controller.go#L1711 and https://github.com/kubernetes/kubernetes/blob/v1.13.0/pkg/volume/plugins.go#L596

**Environment**
- Kubernetes version (use `kubectl version`):
`v1.13.0`
- Driver version:
Latest master build

**Comment by leakingtapan** [comment]
Im able to reproduce the issue too. And it works when in-tree EBS plugin is used with following manifest:
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-test3
spec:
  accessModes:
    - ReadWriteOnce
  capacity:
    storage: 1Gi
  awsElasticBlockStore:
      volumeID: vol-09c88c953e2bb0081
      #fsType: ext3
  persistentVolumeReclaimPolicy: Delete
  volumeMode: Filesystem
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: claim2
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: ""
  resources:
    requests:
      storage: 1Gi
---
apiVersion: v1
kind: Pod
metadata:
  name: app
spec:
  containers:
  - name: app
    image: centos:7
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo $(date -u) >> /data/out.txt; sleep 5; done"]
    volumeMounts:
    - name: persistent-storage
      mountPath: /data
  volumes:
  - name: persistent-storage
    persistentVolumeClaim:
      claimName: claim2
```

I have to comment out `fsType` since there has another issue when it is specified with following error:
```
Warning  FailedMount             5m34s  kubelet, ip-172-20-38-225.ec2.internal  MountVolume.MountDevice failed for volume "pv-test3" : failed to mount the volume as "ext3", it already c
ontains ext4. Mount error: mount failed: exit status 32
Mounting command: systemd-run
Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/mounts/vol-09c88c953e2bb0081 --scope -- mount -t ext3 -o defaults /dev/xvdc
h /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/mounts/vol-09c88c953e2bb0081
Output: Running scope as unit: run-r01d91ea5eb6247fda793b4f38809ba33.scope
mount: wrong fs type, bad option, bad superblock on /dev/xvdch,
```

**Comment by leakingtapan** [comment]
Talked to @msau42 

This is because provisioner annotation is missing from the PV spec, PV deletion work after added annotation

```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-test2
  annotations:
    pv.kubernetes.io/provisioned-by: ebs.csi.aws.com
spec:
  accessModes:
    - ReadWriteOnce
  capacity:
    storage: 1Gi
  csi:
    driver: ebs.csi.aws.com
    fsType: ext3
    volumeHandle: vol-05cee971d935278ac
  persistentVolumeReclaimPolicy: Delete
  volumeMode: Filesystem
```

**Comment by leakingtapan** [comment]
Feature request to external provisioner: https://github.com/kubernetes-csi/external-provisioner/issues/202

**Comment by dkoshkin** [comment]
I wasn't able to get the volume to delete even with the annotation, I'll try with a fresh cluster in a couple of days and report back.

**Comment by leakingtapan** [comment]
I have to make the zone match first, once the app pod is created successfully, the deletion works. Otherwise, if zone mismatches, I got another issue. Do you have the same problem?

**Comment by dkoshkin** [comment]
OK I tried it again on a new Kubernetes `v1.13.1` cluster and the volume is not being deleted:
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-test-pre
  annotations:
    pv.kubernetes.io/provisioned-by: ebs.csi.aws.com
spec:
  accessModes:
    - ReadWriteOnce
  capacity:
    storage: 1Gi
  csi:
    driver: ebs.csi.aws.com
    fsType: ext3
    volumeHandle: vol-0f9431870085724f9
  persistentVolumeReclaimPolicy: Delete
  volumeMode: Filesystem
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: claim-pre
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: ""
  resources:
    requests:
      storage: 1Gi
---
apiVersion: v1
kind: Pod
metadata:
  name: app-pre
spec:
  containers:
  - name: app-pre
    image: centos:7
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo $(date -u) >> /data/out.txt; sleep 5; done"]
    volumeMounts:
    - name: persistent-storage
      mountPath: /data
  volumes:
  - name: persistent-storage
    persistentVolumeClaim:
      claimName: claim-pre
```
Here is some terminal output, you can see that the pod is running and writing out to the volume:
```
➜  aws-ebs-csi-driver git:(e2e-tests) ✗ k get pods
NAME                READY   STATUS    RESTARTS   AGE
app-pre             1/1     Running   0          18s
csi-attacher-0      2/2     Running   0          3m15s
csi-node-n85r5      2/2     Running   0          3m15s
csi-provisioner-0   2/2     Running   0          3m14s
➜  aws-ebs-csi-driver git:(e2e-tests) ✗ k exec -it app-pre cat /data/out.txt
Mon Jan 7 14:48:39 UTC 2019
Mon Jan 7 14:48:44 UTC 2019
➜  aws-ebs-csi-driver git:(e2e-tests) ✗ k get pv
NAME          CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM               STORAGECLASS   REASON   AGE
pv-test-pre   1Gi        RWO            Delete           Bound    default/claim-pre                           38s
➜  aws-ebs-csi-driver git:(e2e-tests) ✗ k get pv pv-test-pre -o yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"PersistentVolume","metadata":{"annotations":{"pv.kubernetes.io/provisioned-by":"ebs.csi.aws.com"},"name":"pv-test-pre"},"spec":{"accessModes":["ReadWriteOnce"],"capacity":{"storage":"1Gi"},"csi":{"driver":"ebs.csi.aws.com","fsType":"ext3","volumeHandle":"vol-0f9431870085724f9"},"persistentVolumeReclaimPolicy":"Delete","volumeMode":"Filesystem"}}
    pv.kubernetes.io/bound-by-controller: "yes"
    pv.kubernetes.io/provisioned-by: ebs.csi.aws.com
  creationTimestamp: "2019-01-07T14:48:23Z"
  finalizers:
  - kubernetes.io/pv-protection
  - external-attacher/ebs-csi-aws-com
  name: pv-test-pre
  resourceVersion: "1521"
  selfLink: /api/v1/persistentvolumes/pv-test-pre
  uid: 4847ee5f-128b-11e9-a86e-0eb0abb42d29
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: claim-pre
    namespace: default
    resourceVersion: "1510"
    uid: 4862c805-128b-11e9-a86e-0eb0abb42d29
  csi:
    driver: ebs.csi.aws.com
    fsType: ext3
    volumeHandle: vol-0f9431870085724f9
  persistentVolumeReclaimPolicy: Delete
  volumeMode: Filesystem
status:
  phase: Bound
```
The controller-manager detaches the volume correctly, and I no longer see the log message about not being able to find the provisioner:
```
I0107 14:48:23.771604       1 reconciler.go:289] attacherDetacher.AttachVolume started for volume "pv-test-pre" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-0f9431870085724f9") from node "kube-node-0-kubelet.devkubernetes01.mesos"
I0107 14:48:27.605627       1 operation_generator.go:335] AttachVolume.Attach succeeded for volume "pv-test-pre" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-0f9431870085724f9") from node "kube-node-0-kubelet.devkubernetes01.mesos"
I0107 14:48:27.605791       1 event.go:221] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"app-pre", UID:"487eb29b-128b-11e9-a86e-0eb0abb42d29", APIVersion:"v1", ResourceVersion:"1517", FieldPath:""}): type: 'Normal' reason: 'SuccessfulAttachVolume' AttachVolume.Attach succeeded for volume "pv-test-pre"
W0107 14:49:57.487295       1 plugins.go:845] FindExpandablePluginBySpec(pv-test-pre) -> err:no volume plugin matched
I0107 14:50:24.728004       1 reconciler.go:230] attacherDetacher.DetachVolume started for volume "pv-test-pre" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-0f9431870085724f9") on node "kube-node-0-kubelet.devkubernetes01.mesos"
I0107 14:50:24.730129       1 operation_generator.go:1243] Verified volume is safe to detach for volume "pv-test-pre" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-0f9431870085724f9") on node "kube-node-0-kubelet.devkubernetes01.mesos"
I0107 14:50:31.586225       1 operation_generator.go:419] DetachVolume.Detach succeeded for volume "pv-test-pre" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-0f9431870085724f9") on node "kube-node-0-kubelet.devkubernetes01.mesos"
```
However the `csi-provisioner` is not getting the message to delete volume:
```
➜  aws-ebs-csi-driver git:(e2e-tests) ✗ k logs csi-provisioner-0 csi-provisioner
I0107 14:45:37.108830       1 csi-provisioner.go:82] Version: v1.0.1-0-g9a258a0a
I0107 14:45:37.108899       1 csi-provisioner.go:96] Building kube configs for running in cluster...
I0107 14:45:37.119166       1 controller.go:197] Connecting to /var/lib/csi/sockets/pluginproxy/csi.sock
I0107 14:45:37.119360       1 controller.go:224] Still trying, connection is CONNECTING
I0107 14:45:37.119605       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:38.119719       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:39.287308       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:40.096783       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:41.234788       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:42.429096       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:43.584980       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:44.649844       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:45.496428       1 controller.go:224] Still trying, connection is CONNECTING
I0107 14:45:45.496474       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:46.608753       1 controller.go:224] Still trying, connection is CONNECTING
I0107 14:45:46.609051       1 controller.go:224] Still trying, connection is TRANSIENT_FAILURE
I0107 14:45:47.119453       1 controller.go:217] Connection timed out
I0107 14:45:57.119593       1 controller.go:197] Connecting to /var/lib/csi/sockets/pluginproxy/csi.sock
I0107 14:45:57.119806       1 controller.go:224] Still trying, connection is CONNECTING
I0107 14:45:57.120021       1 controller.go:221] Connected
I0107 14:45:57.120800       1 controller.go:571] Starting provisioner controller ebs.csi.aws.com_csi-provisioner-0_f118577e-128a-11e9-b7b9-fa26c8dee90b!
I0107 14:45:57.221027       1 controller.go:620] Started provisioner controller ebs.csi.aws.com_csi-provisioner-0_f118577e-128a-11e9-b7b9-fa26c8dee90b!
```
```
➜  aws-ebs-csi-driver git:(e2e-tests) ✗ k logs csi-provisioner-0  ebs-plugin
I0107 14:45:48.735892       1 driver.go:47] Driver: ebs.csi.aws.com Version: latest
I0107 14:45:48.739911       1 mount_linux.go:174] Cannot run systemd-run, assuming non-systemd OS
I0107 14:45:48.739931       1 mount_linux.go:175] systemd-run failed with: exit status 1
I0107 14:45:48.739941       1 mount_linux.go:176] systemd-run output: Failed to create bus connection: No such file or directory
I0107 14:45:48.740105       1 driver.go:90] Listening for connections on address: &net.UnixAddr{Name:"/var/lib/csi/sockets/pluginproxy/csi.sock", Net:"unix"}
```

**Comment by leakingtapan** [comment]
@dkoshkin How did you delete the pod and pvc after app pod is up and running?

**Comment by dkoshkin** [comment]
@leakingtapan interesting I was deleting 3 different files in a directory at the same and it was not deleting the EBS volume.
I did see the same issue about the PV getting stuck being deleted if there was an issue when the pod tries to attach to the volume (I ran into this when the EBS volume was provisioned in a wrong AZ and had to delete the  `volumeattachment` resource to delete the PV)

**Comment by leakingtapan** [comment]
> although the pod deleted succesfully without having to delete the `volumeattachment`

That's what I saw last time I tried, but its not happening for me today (same cluster is used) 😓 

To summarize, there are 2 issues related to this reported buggy behavior:

1. kubectl behaves differently between deleting multiple objects at the same time (either through many files under a directory or one manifest file that contains many specs) and deleting objects individually. I suspect, for the batch deletion, kubectl doesn't check the dependencies between objects. Then, in our case, when pv is deleted before pvc, volume deletion will not be triggered.
2. volumeattachment is not deleted properly, and this causes pv deletion to stuck, two situations could cause this to happen. This seems the same to https://github.com/kubernetes/kubernetes/issues/69697
    1. when volume is scheduled to a mismatched zone
    2. when doing batch deletion using kubectl


@dkoshkin I feel those are upstream k/k issues. Mind creating issues there for the first issue?
